// @peiyanlu/cli-utils@0.0.1/dist/esm/index.mjs
import { exec, spawnSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { copyFile, mkdir, readFile, readdir, rm, writeFile } from "node:fs/promises";
import { join, resolve } from "node:path";

//#region src/enums.ts
let PkgManager = /* @__PURE__ */ function(PkgManager$1) {
	PkgManager$1["NPM"] = "npm";
	PkgManager$1["YARN"] = "yarn";
	PkgManager$1["PNPM"] = "pnpm";
	return PkgManager$1;
}({});
/**
* @deprecated Use `ConfirmResult` instead.
*/
let YesOrNo = /* @__PURE__ */ function(YesOrNo$1) {
	YesOrNo$1["Yes"] = "yes";
	YesOrNo$1["No"] = "no";
	YesOrNo$1["Ignore"] = "ignore";
	return YesOrNo$1;
}({});
let ConfirmResult = /* @__PURE__ */ function(ConfirmResult$1) {
	ConfirmResult$1["YES"] = "yes";
	ConfirmResult$1["NO"] = "no";
	ConfirmResult$1["IGNORE"] = "ignore";
	return ConfirmResult$1;
}({});
let HttpLibrary = /* @__PURE__ */ function(HttpLibrary$1) {
	HttpLibrary$1["EXPRESS"] = "express";
	HttpLibrary$1["FASTIFY"] = "fastify";
	HttpLibrary$1["KOA"] = "koa";
	HttpLibrary$1["HONO"] = "hono";
	return HttpLibrary$1;
}({});

//#endregion
//#region src/utils.ts
const execAsync = (cmd) => {
	return new Promise((r) => {
		exec(cmd, (err, stdout) => r(err ? void 0 : stdout.trim()));
	});
};
const pkgVersion = (pkg) => {
	return execAsync(`npm view ${pkg} version`);
};
const checkVersion = async (cmd) => {
	return execAsync(`${cmd} --version`);
};
const isValidPackageName = (packageName) => {
	return /^(?:@[a-z\d\-*~][a-z\d\-*._~]*\/)?[a-z\d\-~][a-z\d\-._~]*$/.test(packageName);
};
const toValidPackageName = (packageName) => packageName.trim().toLowerCase().replace(/\s+/g, "-").replace(/^[._]/, "").replace(/[^a-z\d\-~]+/g, "-");
const toValidProjectName = (projectName) => projectName.trim().replace(/\/+$/g, "");
const emptyDir = async (dir, ignore = []) => {
	if (!existsSync(dir)) return false;
	for (const file of await readdir(dir)) {
		if (ignore.includes(file)) continue;
		await rm(resolve(dir, file), {
			recursive: true,
			force: true
		});
	}
	return true;
};
const isEmpty = async (path, ignore = []) => {
	return (await readdir(path)).filter((f) => !ignore.includes(f)).length === 0;
};
const editFile = async (file, callback) => {
	if (!existsSync(file)) return;
	return writeFile(file, callback(await readFile(file, "utf-8")), "utf-8");
};
const editJsonFile = (file, callback) => {
	return editFile(file, (str) => {
		try {
			const json = JSON.parse(str);
			callback(json);
			return JSON.stringify(json, null, 2);
		} catch (e) {
			console.error(e);
			return str;
		}
	});
};
const isGitRepo = async (dir) => {
	return !!await execAsync(`git -C "${dir ? `./${dir}` : "."}" rev-parse --is-inside-work-tree`);
};
const readSubDirs = async (source, ignore = []) => {
	return (await readdir(source, { withFileTypes: true })).filter((k) => k.isDirectory() && !ignore.includes(k.name)).map((dir) => dir.name);
};
const getGitConfig = (key, global = true) => {
	return execAsync(`git config ${global ? "--global" : ""} ${key}`);
};
const copyDirAsync = async (src, dest, options) => {
	await mkdir(dest, { recursive: true });
	const entries = await readdir(src, { withFileTypes: true });
	for (const entry of entries) {
		const name = entry.name;
		const isDir = entry.isDirectory();
		const { rename = {}, skips = [] } = options;
		const relName = rename[name] ?? name;
		if (skips.some((rule) => rule(name, isDir))) continue;
		const from = join(src, name);
		const to = join(dest, relName);
		if (isDir) await copyDirAsync(from, to, options);
		else await copyFile(from, to);
	}
};
const readJsonFile = (file) => {
	if (!existsSync(file)) return {};
	try {
		return JSON.parse(readFileSync(file, "utf-8"));
	} catch (e) {
		return {};
	}
};
/**
* 通过包管理器执行脚本时生效
*
* UserAgent: `process.env.npm_config_user_agent`
* @param {string | undefined} userAgent
* @returns {PkgInfo | undefined}
*/
const pkgFromUserAgent = (userAgent) => {
	if (!userAgent) return void 0;
	const [name, version] = userAgent.split(" ")[0].split("/");
	return {
		name,
		version
	};
};
const runCli = (path, args, options) => {
	return spawnSync("node", [path, ...args], {
		env: {
			...process.env,
			_VITE_TEST_CLI: "true"
		},
		encoding: "utf-8",
		...options
	});
};

//#endregion
export { ConfirmResult, HttpLibrary, PkgManager, YesOrNo, checkVersion, copyDirAsync, editFile, editJsonFile, emptyDir, execAsync, getGitConfig, isEmpty, isGitRepo, isValidPackageName, pkgFromUserAgent, pkgVersion, readJsonFile, readSubDirs, runCli, toValidPackageName, toValidProjectName };
